src/Servant/Client_.hs

```diff
@@ -0,0 +1,118 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE DeriveFunctor #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+
+-- |
+-- Module: Servant.Client_
+-- Copyright: Copyright © 2019 Kadena LLC.
+-- License: MIT
+-- Maintainer: Lars Kuhtz <lars@kadena.io>
+-- Stability: experimental
+--
+-- TODO
+--
+module Servant.Client_
+( ClientM_
+, client_
+, runClientM_
+, modifyRequest
+, modifyResponse
+) where
+
+import Control.Lens
+import Control.Monad.Catch
+import Control.Monad.Error.Class
+import Control.Monad.Reader
+
+import Data.Proxy
+
+import GHC.Generics
+
+import Servant.Client
+import Servant.Client.Core
+
+#if ! MIN_VERSION_servant_server(0,16,0)
+type ClientError = ServantError
+#endif
+
+-- -------------------------------------------------------------------------- --
+-- Run Modified Client
+
+data Mods = Mods
+    { _modReq :: !(Request -> IO Request)
+    , _modRes :: !(Response -> IO Response)
+    }
+
+makeLenses ''Mods
+
+-- | 'ClientM_' mimics the behavior of 'ClientM', but provides the functions
+-- 'modifyResponse' and 'modifyRequest' that allow to locally overwrite requests
+-- and responses.
+--
+newtype ClientM_ a = ClientM_ { unClientM_ :: ReaderT Mods ClientM a }
+  deriving
+    ( Functor, Applicative, Monad, MonadIO
+    , Generic
+    , MonadError ClientError, MonadThrow , MonadCatch
+    )
+
+instance MonadReader ClientEnv ClientM_ where
+    ask = ClientM_ $ lift ask
+    local f = ClientM_ . mapReaderT (local f) . unClientM_
+    {-# INLINE ask #-}
+    {-# INLINE local #-}
```

Comment:
isn't this the derived instance for `ClientM_`?

Reply:
`ClientM_` implements `MonadReader ClientEnv`, i.e. it exposes the environment of the underlying `ClientM` monad. 

The fact that `ClientEnv_` itself is a `ReaderT Mods` is a hidden implementation detail and not exported.